30mins-become-contributor-of-tikv.md

```diff
@@ -0,0 +1,276 @@
+
+3. 函数返回值，可以参考 TiDB 中实现的 Eval 函数，对应关系如下：
+
+![表 1.png](https://upload-images.jianshu.io/upload_images/542677-1a493001f3c401ba.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
+
+可以看到 TiDB 的 builtinArithmeticMultiplyIntUnsignedSig  对象实现了 evalInt 方法，故当前函数（`multiply_int_unsigned`）的返回类型应该为 `Result<Option<i64>>`。
+
+4. 函数的参数, 所有 builtin-in 的参数都与 Expression 的 eval 函数一致，即：
+
+	* 环境配置量 (ctx:&StatementContext)
+
+	* 该行数据每列具体值 ( row:&[Datum])
+
+综上，`multiply_int_unsigned` 的下推函数定义为：
+
+```
+   pub fn multiply_int_unsigned(
+
+      &self,
+
+      ctx: &mut EvalContext,
+
+      row: &[Datum],
+
+  ) -> Result<Option<i64>>
+```
+
+
+### Step 4: 实现函数逻辑
+
+这一块相对简单，直接对照 TiDB 的相关逻辑实现即可。这里，我们可以看到 TiDB 的 builtinArithmeticMultiplyIntUnsignedSig  的 具体实现如下：
+
+```
+func (s *builtinArithmeticMultiplyIntUnsignedSig) evalInt(row types.Row) (val int64, isNull bool, err error) {
+
+ a, isNull, err := s.args[0].EvalInt(s.ctx, row)
+
+ if isNull || err != nil {
+
+    return 0, isNull, errors.Trace(err)
+
+ }
+
+ unsignedA := uint64(a)
+
+ b, isNull, err := s.args[1].EvalInt(s.ctx, row)
+
+ if isNull || err != nil {
+
+    return 0, isNull, errors.Trace(err)
+
+ }
+
+ unsignedB := uint64(b)
+
+ result := unsignedA * unsignedB
+
+ if unsignedA != 0 && result/unsignedA != unsignedB {
+
+    return 0, true, types.ErrOverflow.GenByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s * %s)", s.args[0].String(), s.args[1].String()))
+
+ }
+
+ return int64(result), false, nil
+
+}
+
+```
+
+参考以上代码，翻译到 TiKV 即可，如下：
+
+```
+pub fn multiply_int_unsigned(
+
+      &self,
+
+      ctx: &mut EvalContext,
+
+      row: &[Datum],
+
+  ) -> Result<Option<i64>> {
+
+      let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+
+      let rhs = try_opt!(self.children[1].eval_int(ctx, row));
+
+      let res = (lhs as u64).checked_mul(rhs as u64).map(|t| t as i64);
+
+      // TODO: output expression in error when column's name pushed down.
+
+      res.ok_or_else(|| Error::overflow("BIGINT UNSIGNED", &format!("({} * {})", lhs, rhs)))
+
+          .map(Some)
+
+  }
+```
+
+### Step 5: 添加参数检查
+
+TiKV 在收到下推请求时，首先会对所有的表达式进行检查，表达式的参数个数检查就在这一步进行。
+
+TiDB 中对每个 builtin 函数的参数个数有严格的限制，这一部分检查可参考 TiDB 同目录下 builtin.go 相关代码。
+
+在 TiKV 同级目录的 `scalar_function.rs` 文件里，找到 ScalarFunc 的 `check_args` 函数，按照现有的模式，加入参数个数的检查即可。
+
+### Step 6: 添加下推支持
+
+TiKV 在对一行数据执行具体的 expression 时，会调用 eval 函数，eval 函数又会根据具体的返回类型，执行具体的子函数。这一部分工作在 `scalar_function.rs` 中以宏（dispatch_call）的形式完成。
+
+对于 MultiplyIntUnsigned, 我们最终返回的数据类型为 Int，所以可以在 dispatch_call 中找到 `INT_CALLS`，然后照着加入 `MultiplyIntUnsigned => multiply_int_unsigned` , 表示当解析到函数签名 MultiplyIntUnsigned 时，调用上述已实现的函数 `multiply_int_unsigned`。
+
+至此 MultiplyIntUnsigned 下推逻辑已完全实现。
+
+### Step 7: 添加测试
+
+在函数 `multiply_int_unsigned` 所在文件 [builtin_arithmetic.rs](https://github.com/pingcap/tikv/blob/master/src/coprocessor/dag/expr/builtin_arithmetic.rs) 底部的 test 模块中加入对该函数签名的单元测试，要求覆盖到上述添加的所有代码，这一部分也可以参考 TiDB 中相关的测试代码。本例在 TiKV 中实现的测试代码如下：
+
+```
+   #[test]
```

Comment:
Please copy it from source https://github.com/pingcap/tikv/blob/master/src/coprocessor/dag/expr/builtin_arithmetic.rs#L954